src/posts/en/temporal-dead-zone-tdz-demystified.md

```diff
@@ -0,0 +1,193 @@
+<!--
+layout: post
+title: Temporal Dead Zone (TDZ) demystified
+date: 2015-01-29T01:18:37.630Z
+comments: true
+published: true
+keywords: scope, tdz, es6
+description: Get a deeper understanding of scopes and future-proof your code!
+categories: scope, articles
+authorName: Fabrício S. Matté
+authorLink: https://twitter.com/UltCombo
+authorDescription: ECMAScript enthusiast, open source addict and Web Platform lover.
+authorPicture: https://pbs.twimg.com/profile_images/490627147963187200/2BiH3pv4.png
+-->
+<!--more-->
+The Temporal Dead Zone refers to a new set of ECMAScript semantics regarding scope, introduced in ES2015 (aka ES6).
+
+Although the name may sound a bit intimidating, the concept is not hard to grasp, actually. But first, let's take one step back and look at how scoping works in ES5:
+
+```javascript
+var x = 'outer scope';
+(function() {
+    console.log(x);
+    var x = 'inner scope';
+}());
+```
+
+Upon executing the code above, can you tell what the `console.log(x)` prints? If you guessed `undefined`, then keep reading. Otherwise, please take the time to read about declarations hoisting ([MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/var#var_hoisting), [Adequately Good](http://www.adequatelygood.com/JavaScript-Scoping-and-Hoisting.html)) and [variable shadowing](http://en.wikipedia.org/wiki/Variable_shadowing), as these concepts are key to fully understanding TDZ.
+
+---
+
+# Enter the Temporal Dead Zone
+
+Alright, now let's take a small step forward and begin with an extremely simple and contrived TDZ scoping example:
+
+```javascript
+console.log(x); // throws a ReferenceError
+let x = 'hey';
+```
+
+As you can see, one of the main differences between the old `var` and the new `let`/`const` declarations (besides their [scope](/2014/08/what-you-need-to-know-about-block-scope-let/)) is that the latter are constrained by the Temporal Dead Zone semantics, meaning they will throw a `ReferenceError` when accessed (read/write) before being initialized, instead of returning `undefined` as a `var`-declared variable would. This makes the code more predictable and easier to spot potential bugs. Simple, isn't it?
+
+# Well, TDZ is not quite so simple
+
+Taking a second look at the previous example, one could deduce that `let`/`const` declarations simply do not hoist, and that would explain the `ReferenceError`, right? **Nope**, that's an incorrect over-simplification *(and beware of uninformed resources claiming that!)*.
+
+Let's go back to the very first example in this article, replacing `var` with `let` and see what happens:
+
+```javascript
+let x = 'outer scope';
+(function() {
+    console.log(x);
+    let x = 'inner scope';
+}());
```

Comment:
I think this would benefit from having a block version of the example: 

```js
let x = 'outer';

{
  console.log(x);
  var x = 'inner scope';
}
```